### PR TITLE
Add row pitch to Buffer->Texture and Texture->Buffer copies

### PR DESCRIPTION
**Pull request by austinEng**

### CODE REVIEW

File: src/tests/end2end/InputStateTests.cpp [hunk 1]
```diff
@@ -384,11 +384,6 @@ TEST_P(InputStateTest, TwoAttributesAtAnOffsetInstance) {
 
 // Test a pure-instance input state
 TEST_P(InputStateTest, PureInstance) {
-    if (IsD3D12()) {
```
Comment (Kangz) [inline src/tests/end2end/InputStateTests.cpp]: I think this patch needs to be rebased, I've already landed this change.

File: src/tests/unittests/validation/CopyCommandsValidationTests.cpp [hunk 2]
```diff
@@ -114,89 +114,110 @@ class CopyCommandTest_B2T : public CopyCommandTest {
 
 // Test a successfull B2T copy
 TEST_F(CopyCommandTest_B2T, Success) {
-    nxt::Buffer source = CreateFrozenBuffer(16 * 4, nxt::BufferUsageBit::TransferSrc);
-    nxt::Texture destination = CreateFrozen2DTexture(16, 16, 5, nxt::TextureFormat::R8G8B8A8Unorm,
+    nxt::Buffer source = CreateFrozenBuffer(256 * 4, nxt::BufferUsageBit::TransferSrc);
+    nxt::Texture destination = CreateFrozen2DTexture(256, 16, 5, nxt::TextureFormat::R8G8B8A8Unorm,
```
Comment (Kangz) [inline src/tests/unittests/validation/CopyCommandsValidationTests.cpp]: The size of the texture doesn't need to be changed, only the buffer needs to have enough space for the extra padding. All of the texture related arguments below can stay to their original values and rowPitch can be left to 0 in most cases.

File: src/tests/unittests/validation/CopyCommandsValidationTests.cpp [hunk 3]
```diff
@@ -114,89 +114,110 @@ class CopyCommandTest_B2T : public CopyCommandTest {
...
             // Copy with a buffer offset
-            .CopyBufferToTexture(source, 15 * 4, destination, 0, 0, 0, 1, 1, 1, 4)
+            .CopyBufferToTexture(source, 192 * 4, 0, destination, 0, 0, 0, 64, 1, 1, 2)
+            .GetResult();
+    }
+
+    // Copies with a 256-byte aligned row pitch but unaligned texture region
+    {
+        nxt::CommandBuffer commands = AssertWillBeSuccess(device.CreateCommandBufferBuilder())
+            // Unaligned block, aligned row pitch
```
Comment (Kangz) [inline src/tests/unittests/validation/CopyCommandsValidationTests.cpp]: nit: the comment above mentions "region" but this one mentions "block". Also the rowpitch is always aligned in this test so we might avoid mentioning it.
Also: nit: for misalignment in the texture can we use even more unaligned numbers like 5, 17, 23 and friends? (The buffer offset should still be aligned at a texel size though.)

File: src/tests/unittests/validation/CopyCommandsValidationTests.cpp [hunk 4]
```diff
@@ -114,89 +114,110 @@ class CopyCommandTest_B2T : public CopyCommandTest {
...
-            .CopyBufferToTexture(source, 0, destination, 0, 0, 0, 4, 4, 1, 2)
+            // Copy 64x4 block in corner of first mip.
+            .CopyBufferToTexture(source, 0, 0, destination, 0, 0, 0, 64, 4, 1, 0)
+            // Copy 64x4 block in opposite corner of first mip.
+            .CopyBufferToTexture(source, 0, 0, destination, 192, 12, 0, 64, 4, 1, 0)
+            // Copy 64x4 block in the 64x4 mip.
+            .CopyBufferToTexture(source, 0, 0, destination, 0, 0, 0, 64, 4, 1, 2)
             // Copy with a buffer offset
-            .CopyBufferToTexture(source, 15 * 4, destination, 0, 0, 0, 1, 1, 1, 4)
+            .CopyBufferToTexture(source, 192 * 4, 0, destination, 0, 0, 0, 64, 1, 1, 2)
```
Comment (Kangz) [inline src/tests/unittests/validation/CopyCommandsValidationTests.cpp]: nit: we could do this for the 0th mip.

File: src/tests/unittests/validation/CopyCommandsValidationTests.cpp [hunk 5]
```diff
@@ -114,89 +114,110 @@ class CopyCommandTest_B2T : public CopyCommandTest {
...
+
+    // Copies with a 256-byte aligned row pitch but unaligned texture region
+    {
+        nxt::CommandBuffer commands = AssertWillBeSuccess(device.CreateCommandBufferBuilder())
+            // Unaligned block, aligned row pitch
+            .CopyBufferToTexture(source, 0, 256, destination, 0, 0, 0, 32, 4, 1, 0)
+            // Unaligned offset block, aligned row pitch
+            .CopyBufferToTexture(source, 0, 256, destination, 192, 12, 0, 32, 4, 1, 0)
+            // Unaligned block, row pitch > width
+            .CopyBufferToTexture(source, 0, 512, destination, 0, 0, 0, 64, 2, 1, 2)
```
Comment (Kangz) [inline src/tests/unittests/validation/CopyCommandsValidationTests.cpp]: nit: use the 0th mip when it isn't to test it specifically.

File: src/tests/unittests/validation/CopyCommandsValidationTests.cpp [hunk 6]
```diff
@@ -114,89 +114,110 @@ class CopyCommandTest_B2T : public CopyCommandTest {
...
+            .CopyBufferToTexture(source, 0, 0, destination, 256, 16, 0, 0, 0, 1, 0)
             .GetResult();
     }
 }
 
 // Test OOB conditions on the buffer
 TEST_F(CopyCommandTest_B2T, OutOfBoundsOnBuffer) {
-    nxt::Buffer source = CreateFrozenBuffer(16 * 4, nxt::BufferUsageBit::TransferSrc);
-    nxt::Texture destination = CreateFrozen2DTexture(16, 16, 5, nxt::TextureFormat::R8G8B8A8Unorm,
+    nxt::Buffer source = CreateFrozenBuffer(256 * 4, nxt::BufferUsageBit::TransferSrc);
```
Comment (Kangz) [inline src/tests/unittests/validation/CopyCommandsValidationTests.cpp]: Same comments here and in other tests as in the success test.

File: src/tests/unittests/validation/CopyCommandsValidationTests.cpp [hunk 7]
```diff
@@ -114,89 +114,110 @@ class CopyCommandTest_B2T : public CopyCommandTest {
...
 
     // OOB on the texture even on an empty copy when we copy to a non-existent mip.
     {
         nxt::CommandBuffer commands = AssertWillBeError(device.CreateCommandBufferBuilder())
-            .CopyBufferToTexture(source, 0, destination, 0, 0, 0, 0, 0, 1, 5)
+            .CopyBufferToTexture(source, 0, 0, destination, 0, 0, 0, 0, 0, 1, 5)
+            .GetResult();
+    }
+
+    // OOB on the texture because row pitch * height overlows
```
Comment (Kangz) [inline src/tests/unittests/validation/CopyCommandsValidationTests.cpp]: typo: overflows

File: src/tests/unittests/validation/CopyCommandsValidationTests.cpp [hunk 8]
```diff
@@ -114,89 +114,110 @@ class CopyCommandTest_B2T : public CopyCommandTest {
...
-            .CopyBufferToTexture(source, 0, destination, 0, 0, 0, 4, 5, 1, 0)
+            .CopyBufferToTexture(source, 0, 0, destination, 0, 0, 0, 64, 5, 1, 0)
             .GetResult();
     }
 
     // OOB on the buffer because of the offset
     {
         nxt::CommandBuffer commands = AssertWillBeError(device.CreateCommandBufferBuilder())
-            .CopyBufferToTexture(source, 1, destination, 0, 0, 0, 4, 4, 1, 0)
+            .CopyBufferToTexture(source, 1, 0, destination, 0, 0, 0, 64, 4, 1, 0)
```
Comment (Kangz) [inline src/tests/unittests/validation/CopyCommandsValidationTests.cpp]: nit: can you make the buffer offset be 4 here and in the respective T2B test?

File: src/backend/CommandBuffer.cpp [hunk 9]
```diff
@@ -69,10 +69,23 @@ namespace backend {
             return true;
         }
 
-        bool ComputeTextureCopyBufferSize(CommandBufferBuilder*, const TextureCopyLocation& location, uint32_t* bufferSize) {
+        bool ComputeTextureCopyBufferSize(CommandBufferBuilder*, const TextureCopyLocation& location, uint32_t rowPitch, uint32_t* bufferSize) {
             // TODO(cwallez@chromium.org): check for overflows
-            uint32_t pixelSize = static_cast<uint32_t>(TextureFormatPixelSize(location.texture->GetFormat()));
-            *bufferSize = location.width * location.height * location.depth * pixelSize;
+            *bufferSize = rowPitch * location.height * location.depth;
```
Comment (Kangz) [inline src/backend/CommandBuffer.cpp]: Depending on Ben's answer to the D3D12 placed footprint thread, this could be ((height - 1) * rowPitch + width) * depth. We would need to add a test for this corner case too.

File: src/backend/CommandBuffer.cpp [hunk 10]
```diff
@@ -69,10 +69,23 @@ namespace backend {
...
+
+            return true;
+        }
+
+        uint32_t ComputeDefaultRowPitch(TextureBase* texture, uint32_t width) {
```
Comment (Kangz) [inline src/backend/CommandBuffer.cpp]: Do we want to make this align the rowPitch on its own?

File: src/tests/end2end/InputStateTests.cpp (hunk 1 above)
Comment (austinEng) [inline src/tests/end2end/InputStateTests.cpp]: I think I'm up to date. The test was fixed but it wasn't re-enabled.

File: src/tests/end2end/BasicTests.cpp [hunk 11]
```diff
@@ -35,8 +35,9 @@ TEST_P(BasicTests, BufferSetSubData) {
 }
 
 TEST_P(BasicTests, ReadPixelsTest) {
-    RGBA8 red(255, 0, 0, 255);
-    nxt::Buffer buffer = utils::CreateFrozenBufferFromData(device, &red, sizeof(red), nxt::BufferUsageBit::TransferSrc);
+    RGBA8 data[64];
+    data[0] = RGBA8(255, 0, 0, 255);
+    nxt::Buffer buffer = utils::CreateFrozenBufferFromData(device, data, sizeof(data), nxt::BufferUsageBit::TransferSrc);
```
Comment (austinEng) [inline src/tests/end2end/BasicTests.cpp]: I don't know about the other backends, but D3D12 requires the buffer to be at least as large as the placed footprint. In validation, `ComputeTextureCopyBufferSize` now computes the buffer size as `rowPitch * location.height * location.depth`

File: src/backend/CommandBuffer.cpp (hunk 9 above)
Comment (austinEng) [inline src/backend/CommandBuffer.cpp]: If I remember correctly, I do think this is what the their helper function for computing the placed footprint does.

File: src/backend/CommandBuffer.cpp (hunk 10 above)
Comment (austinEng) [inline src/backend/CommandBuffer.cpp]: I think that if the application sets the rowPitch to be 0, they intend to mean that the data is tightly packed and not that NXT should automatically do something that "works".

File: src/tests/end2end/BasicTests.cpp (hunk 11 above)
Comment (austinEng) [inline src/tests/end2end/BasicTests.cpp]: Nevermind. I realized if we do `((height - 1) * rowPitch + width) * depth` then this is fine

File: src/backend/CommandBuffer.cpp (hunk 10 above)
Comment (Kangz) [inline src/backend/CommandBuffer.cpp]: Sounds good

File: src/tests/end2end/InputStateTests.cpp (hunk 1 above)
Comment (Kangz) [inline src/tests/end2end/InputStateTests.cpp]: Ooops, I should have done it :)

File: src/tests/unittests/validation/CopyCommandsValidationTests.cpp (hunk 2 above)
Comment (austinEng) [inline src/tests/unittests/validation/CopyCommandsValidationTests.cpp]: I think we'll need to update rowPitch to be 256 in most cases. A 4x4 region needs a rowPitch of 256 and not 16

File: src/backend/CommandBuffer.cpp [hunk 12]
```diff
@@ -69,10 +69,23 @@ namespace backend {
...
+            uint32_t pixelSize = static_cast<uint32_t>(TextureFormatPixelSize(texture->GetFormat()));
+            return pixelSize * width;
+        }
+
+        bool ValidateRowPitch(CommandBufferBuilder* builder, uint32_t rowPitch) {
+            if (rowPitch % 256 != 0) {
```
Comment (austinEng) [inline src/backend/CommandBuffer.cpp]: Done

File: src/backend/CommandBuffer.cpp [hunk 13]
```diff
@@ -403,7 +416,8 @@ namespace backend {
                         CopyBufferToTextureCmd* copy = iterator.NextCommand<CopyBufferToTextureCmd>();
 
                         uint32_t bufferCopySize = 0;
-                        if (!ComputeTextureCopyBufferSize(this, copy->destination, &bufferCopySize) ||
+                        if (!ValidateRowPitch(this, copy->rowPitch) ||
```
Comment (austinEng) [inline src/backend/CommandBuffer.cpp]: Done

File: src/tests/NXTTest.cpp [hunk 14]
```diff
@@ -156,14 +156,17 @@ void NXTTest::AddBufferExpectation(const char* file, int line, const nxt::Buffer
...
+    deferred.pitch = size;
     deferred.expectation = expectation;
 
     deferredExpectations.push_back(deferred);
 }
 
 void NXTTest::AddTextureExpectation(const char* file, int line, const nxt::Texture& texture, uint32_t x, uint32_t y, uint32_t width, uint32_t height, uint32_t pixelSize, detail::Expectation* expectation) {
     nxt::Texture source = texture.Clone();
-    uint32_t size = width * height * pixelSize;
+    uint32_t rowPitch = ((width * pixelSize - 1) / 256 + 1) * 256;
```
Comment (austinEng) [inline src/tests/NXTTest.cpp]: Done

File: src/tests/NXTTest.cpp [hunk 15]
```diff
@@ -244,19 +249,41 @@ void NXTTest::SlotMapReadCallback(nxtBufferMapReadStatus status, const void* dat
...
-    for(const auto& expectation : deferredExpectations) {
+    for (const auto& expectation : deferredExpectations) {
         NXT_ASSERT(readbackSlots[expectation.readbackSlot].mappedData != nullptr);
 
         // Get a pointer to the mapped copy of the data for the expectation.
         const char* data = reinterpret_cast<const char*>(readbackSlots[expectation.readbackSlot].mappedData);
         data += expectation.readbackOffset;
 
+        uint32_t size;
+        char* packedData = nullptr;
```
Comment (austinEng) [inline src/tests/NXTTest.cpp]: Done

File: src/tests/NXTTest.h [hunk 16]
```diff
@@ -102,6 +104,8 @@ class NXTTest : public ::testing::TestWithParam<BackendType> {
             size_t readbackSlot;
             uint32_t readbackOffset;
             uint32_t size;
+            uint32_t width;
```
Comment (austinEng) [inline src/tests/NXTTest.h]: Done

File: src/tests/unittests/validation/CopyCommandsValidationTests.cpp (hunk 4 above)
Comment (austinEng) [inline src/tests/unittests/validation/CopyCommandsValidationTests.cpp]: Done

File: src/tests/unittests/validation/CopyCommandsValidationTests.cpp (hunk 3 above)
Comment (austinEng) [inline src/tests/unittests/validation/CopyCommandsValidationTests.cpp]: Done

File: src/tests/unittests/validation/CopyCommandsValidationTests.cpp (hunk 5 above)
Comment (austinEng) [inline src/tests/unittests/validation/CopyCommandsValidationTests.cpp]: Done

File: src/tests/unittests/validation/CopyCommandsValidationTests.cpp (hunk 6 above)
Comment (austinEng) [inline src/tests/unittests/validation/CopyCommandsValidationTests.cpp]: Done

File: src/tests/unittests/validation/CopyCommandsValidationTests.cpp (hunk 8 above)
Comment (austinEng) [inline src/tests/unittests/validation/CopyCommandsValidationTests.cpp]: Done

File: src/tests/unittests/validation/CopyCommandsValidationTests.cpp (hunk 7 above)
Comment (austinEng) [inline src/tests/unittests/validation/CopyCommandsValidationTests.cpp]: Done

File: src/tests/unittests/validation/CopyCommandsValidationTests.cpp [hunk 17]
```diff
@@ -234,14 +274,42 @@ TEST_F(CopyCommandTest_B2T, IncorrectUsage) {
...
+    {
+        nxt::CommandBuffer commands = AssertWillBeError(device.CreateCommandBufferBuilder())
+            .CopyBufferToTexture(source, 0, 0, destination, 0, 0, 0, 3, 4, 1, 0)
+            .GetResult();
+    }
+
+    // Row pitch is not 256-byte aligned
+    {
+        nxt::CommandBuffer commands = AssertWillBeError(device.CreateCommandBufferBuilder())
+            .CopyBufferToTexture(source, 0, 257, destination, 0, 0, 0, 4, 4, 1, 0)
```
Comment (Kangz) [inline src/tests/unittests/validation/CopyCommandsValidationTests.cpp]: Can it be 128? 257 is very not aligned :)

File: src/common/Math.h [hunk 18]
```diff
@@ -27,6 +27,11 @@ bool IsAligned(const void* ptr, size_t alignment);
 void* AlignVoidPtr(void* ptr, size_t alignment);
 
 template<typename T>
+T Align(T value, size_t alignment) {
+    return static_cast<T>(reinterpret_cast<uintptr_t>(AlignVoidPtr(reinterpret_cast<void*>(static_cast<uintptr_t>(value)), alignment)));
```
Comment (austinEng) [inline src/common/Math.h]: I don't think it's valid to have a templated and non-templated function both called the same thing. We could name one AlignNumber or AlignPtr?

Comment (Kangz) [inline src/common/Math.h]: Good point, let's call Align(T*) AlignPtr

Comment (Kangz) [inline src/common/Math.h]: Was done.

File: src/tests/unittests/MathTests.cpp [hunk 19]
```diff
@@ -66,6 +66,14 @@ TEST(Math, Align) {
         ASSERT_LT(static_cast<size_t>(aligned - unaligned), kTestAlignment);
         ASSERT_EQ(reinterpret_cast<intptr_t>(aligned) & (kTestAlignment -1), 0);
     }
+
+    for (uint32_t i = 0; i < 2 * kTestAlignment; ++i) {
```
Comment (Kangz) [inline src/tests/unittests/MathTests.cpp]: Was done.

File: src/tests/unittests/validation/CopyCommandsValidationTests.cpp [hunk 20]
```diff
@@ -37,6 +39,11 @@ class CopyCommandTest : public ValidationTest {
             tex.FreezeUsage(usage);
             return tex;
         }
+
+        uint32_t bufferSizeForTextureCopy(uint32_t width, uint32_t height, uint32_t depth) {
```
Comment (Kangz) [inline src/tests/unittests/validation/CopyCommandsValidationTests.cpp]: Was done.

File: src/tests/unittests/validation/CopyCommandsValidationTests.cpp (hunk 17 above)
Comment (Kangz) [inline src/tests/unittests/validation/CopyCommandsValidationTests.cpp]: Was done.

**Comment by Kangz** [comment]
LGTM with a rebase and handling of rowPitch in CommandBufferGL's TextureToBuffer command handling. Sorry for the merge conflict.

File: src/backend/opengl/CommandBufferGL.cpp [hunk 21]
```diff
@@ -210,9 +212,11 @@ namespace opengl {
                                                texture->GetHandle(), src.level);
 
                         glBindBuffer(GL_PIXEL_PACK_BUFFER, buffer->GetHandle());
+                        glPixelStorei(GL_PACK_ROW_LENGTH, copy->rowPitch / static_cast<uint32_t>(TextureFormatPixelSize(texture->GetFormat())));
```
Comment (Kangz) [inline src/backend/opengl/CommandBufferGL.cpp]: Thanks for handling this one!
